q2_feature_classifier/_consensus_assignment.py

```diff
@@ -0,0 +1,243 @@
+        if output_no_hits is True:
+            consensus = _output_no_hits(
+                consensus, exp_seq_ids, unassignable_label)
+        # if there are no consensus results (no hits for any query), default
+        # to empty result file (must create empty dict or fail on pd error)
+        if not consensus:
+            consensus = {'': ('', '')}
+        result = pd.DataFrame.from_dict(consensus, 'index')
+        result.index.name = 'Feature ID'
+        result.columns = ['Taxon', 'Confidence']
+        return result
+
+
+def _validate_params(min_id, maxaccepts, min_consensus):
+    if min_id > 1.0 or min_id <= 0.0:
+        raise ValueError('min_id must be > 0 and <= 1')
+    if maxaccepts < 1:
+        raise ValueError('maxaccepts must be >= 1')
+    if min_consensus > 1.0 or min_consensus <= 0.5:
+        raise ValueError('min_consensus must be > 0.51 and <= 1.0')
+
+
+def _output_no_hits(obs_taxa, exp_seq_ids,
+                    unassignable_label=_get_default_unassignable_label()):
+    '''If a query ID has no hits, report as unassigned.'''
+    # extract list of query sequence IDs
+    exp = _open_list_or_file(exp_seq_ids)
+    exp = [line.strip('>') for line in exp if line.startswith('>')]
+
+    # iterate over expected IDs, add as unassigned to obs_taxa if missing
+    for _id in exp:
+        if _id not in obs_taxa:
+            obs_taxa[_id] = (unassignable_label, 0.0)
+
+    return obs_taxa
+
+
+def _run_command(cmd, verbose=True):
+    if verbose:
+        print("Running external command line application. This may print "
+              "messages to stdout and/or stderr.")
+        print("The command being run is below. This command cannot "
+              "be manually re-run as it will depend on temporary files that "
+              "no longer exist.")
+        print("\nCommand:", end=' ')
+        print(" ".join(cmd), end='\n\n')
+    subprocess.run(cmd, check=True)
+
+
+def _import_blast_assignments(
+        assignments, ref_taxa,
+        unassignable_label=_get_default_unassignable_label()):
+    '''import observed assignments to dict of lists.
+
+    assignments: path or list
+        Taxonomy observation map in blast format 6 or 7. Each line consists of
+        taxonomy assignments of a query sequence in tab-delimited format:
+            <query_id>    <assignment_id>   <...other columns are ignored>
+
+    ref_taxa: dict or pd.Series
+        Reference taxonomies in tab-delimited format:
+            <accession ID>  kingdom;phylum;class;order;family;genus;species
+    '''
+    obs_taxa = {}
+    lines = _open_list_or_file(assignments)
+
+    for line in lines:
+        if not line.startswith('#') or line == "":
+            i = line.split('\t')
+            # ref taxonomy IDs get imported as a str when using CLI
+            # but imported as int in python interpreter (e.g., during tests)
+            # the following allows dict lookup as str or int.
+            try:
```

Comment:
Using `if`/`else` statements instead of `try`/`except` is typically more clear. You also typically want to use `try`/`expect` for "unusual" cases, and `if`/`else` for cases that frequently come up. You could do this as:

```python
if fields[1] == '*':
    t = [unassignable_label]
else:
    try:
        t = ref_taxa[fields[1]].split(';')
    except KeyError:
        raise KeyError('Identifier %s was reported in taxonomic search results, but was not present in the reference taxonomy.' % fields[1])
```

For this to work, you'll need to ensure that the indices in `ref_taxa` are always strings. Is that possible?